pytorch_lightning/metrics/metric.py

```diff
@@ -27,18 +39,90 @@ def __init__(self, name: str):
         """
         super().__init__()
         self.name = name
+        self._dtype = torch.get_default_dtype()
+        self._device = torch.device('cpu')
+
+        # Register hooks
+        self.register_forward_pre_hook(self.input_convert)
+        self.register_forward_hook(self.output_convert)
+        self.register_forward_hook(self.ddp_sync)
+        self.register_forward_hook(self.compute)
 
     @abstractmethod
-    def forward(self, *args, **kwargs) -> torch.Tensor:
+    def forward(self, *args, **kwargs):
         """
         Implements the actual metric computation.
 
         Returns:
-            metric value
+            metric value or metric state
 
         """
         raise NotImplementedError
 
+    def compute(self, module: nn.Module, data: Any, output: Any):
```

Comment:
why do you need the metric module? Shouldn't this be available through `self?`

Reply:
Yes, you are right. A forward hook always takes 3 arguments: the module, input to the forward method and output of the forward method (https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.register_forward_hook). So having both a self and module argument was redundant. I deleted one arg and made these functions as @staticmethods instead